api/auth/external_auth.proto

```diff
@@ -0,0 +1,147 @@
+syntax = "proto3";
+
+// [#proto-status: draft]
+
+package envoy.api.v2.auth;
+
+import "api/address.proto";
+import "google/protobuf/struct.proto";
+import "google/protobuf/timestamp.proto";
+import "google/rpc/status.proto";
+
+// Authorization service.
+// Primarily responds with `OK` or `NOT OK`.
+service Authorization {
+  rpc Check(CheckRequest) returns (CheckResponse);
+}
+
+message CheckRequest {
+  // The request attributes.
+  AttributeContext attributes = 1;
+}
+
+message CheckResponse {
+  // Status `OK` allows the request. Any other status indicates the request should be denied.
+  google.rpc.Status status = 1;
+}
+
+// An attribute is a piece of metadata that describes an activity on a network.
+// For example, the size of an HTTP request, or the status code of an HTTP response.
+//
+// Each attribute has a type and a name, which is logically defined as a proto message field.
+// AttributeContext is a collection of individual attributes.
+message AttributeContext {
+  // This message defines attributes for a node that handles a network request.
+  // The node can be either a service or an application that sends, forwards,
+  // or receives the request. Service peers should fill in the `service`,
+  // `principal`, and `labels` as appropriate.
+  message Peer {
+    // The address of the peer, this is typically the IP address.
+    // It can also be UDS, or others.
+    Address address = 1;
```

Comment:
This change drops source.ip attribute, which is very important for Istio. Can we duplicated the ip at Peer level?

Users will write a lot of configs with reference to "source.ip". With the current change, the attribute becomes "source.address.socket_address.address". That is totally not acceptable. The attribute context is to carry attributes in a user friendly way. The Address proto is for machine processing. I strongly recommend we focus on users, not on machine.

If we need, we can convince both Envoy and Istio community to use string address. And encode "1.1.1.1" or "uds:/tmp/xx" or "tcp:1.1.1.1". Most network traffic carries via IP network. By treating different network protocol equally, we significantly hurt usability. So I am against to use Address here.

Reply:
+1 for duplication of ip.

Reply:
@wora I'm sorry but I disagree that the goal here is to have protos that are user friendly. It really doesn't make any sense to me. I think we are vastly over complicating this PR and the concepts herein in general. This is about sending data that Envoy knows about to a management server which will do "stuff" and then respond. I don't see any scenario in which the data that Envoy sends is going to be consumed directly by a user. Various management servers will offer their own configurations and map the Envoy data on to it. IMO making these protos human friendly and generic is a non-goal.

In this particular case, I would strongly suggest that you do what I have done on the access log side. See this PR: https://github.com/envoyproxy/data-plane-api/pull/344. We should be sending both remote and local address to the auth server.